processor/groupbytraceprocessor/processor.go

```diff
@@ -0,0 +1,343 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package groupbytraceprocessor
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/component/componenterror"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/pdata"
+)
+
+var (
+	errNoInstrumentationLibrarySpans = errors.New("trace didn't contain any instrumentation library spans")
+	errNoSpans                       = errors.New("instrumentation library spans didn't contain any spans")
+	errNilTrace                      = errors.New("invalid trace (nil)")
+	errNilTraceID                    = errors.New("invalid trace ID (nil)")
+)
+
+// groupByTraceProcessor is a processor that keeps traces in memory for a given duration, with the expectation
+// that the trace will be complete once this duration expires. After the duration, the trace is sent to the next consumer.
+// This processor avoid locks in the hot-path by using channels for synchronization and communication.
+// The internal map is changed only by processes running in the main loop, ensuring that no concurrent reads and
+// writes occur (concurrent reads are fine).
+// The storage may be changed by different go routines, requiring the storage implementation to be thread-safe.
+//
+// The typical data flow looks like:
+// ConsumeTraces() -> chan received -> onTrace() -> processBatch() -> processTrace() ->
+// time.AfterFunc(chan expired) -> onTrace() -> processExpired() -> async markAsReleased() ->
+// chan released -> onReleased() -> nextConsumer.ConsumeTraces()
+type groupByTraceProcessor struct {
+	nextConsumer consumer.TraceConsumer
+	config       Config
+	logger       *zap.Logger
+
+	// used for the ring buffer
+	index     int
+	ids       []pdata.TraceID
+	idToIndex map[string]int // key is traceID as string, value is the index on the 'ids' slice
+
+	// the trace storage
+	st storage
+
+	// traces received from the previous processors
+	received chan pdata.Traces
```

Comment:
Would it potentially be easier to read if you had just one channel with "event type" attached to event (received / expired / released / removed), and a single loop processing all of them?
It's hard to keep mental model of juggling with 4 channels.

I see that you are also using different data types for each channel, so not sure if that's feasible.

Alternatively, you may also consider making a "state machine" given the transition graph you have documented above.

Reply:
I like the idea of the state machine. I'll give this a try.

Reply:
I just added another commit to this PR adding an event machine. I think it's now more readable as the concerns are separated. The performance also seems to be a bit better:

```console
$ go test -benchmem -run=^$ go.opentelemetry.io/collector/processor/groupbytraceprocessor -bench "^(BenchmarkConsumeTracesCompleteOnFirstBatch)$"
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/collector/processor/groupbytraceprocessor
BenchmarkConsumeTracesCompleteOnFirstBatch-12    	 1393412	       848 ns/op	    1107 B/op	      18 allocs/op
PASS
ok  	go.opentelemetry.io/collector/processor/groupbytraceprocessor	2.085s
```